rtl/ip/xadc/rtl/xadc_adapter.sv

```
@@ -0,0 +1,153 @@
+    .tl_o        (tl_o),
+
+    .en_ifetch_i (prim_mubi_pkg::MuBi4False),
+
+    .req_o       (mem_req),
+    .req_type_o  (),
+    .gnt_i       (mem_gnt),
+    .we_o        (mem_we),
+    .addr_o      (mem_addr),
+    .wdata_o     (mem_wdata),
+    .wdata_cap_o (),
+    .wmask_o     (mem_wmask),
+    .intg_error_o(),
+    .rdata_i     (mem_rdata),
+    .rdata_cap_i (1'b0),
+    .rvalid_i    (mem_rvalid),
+    .rerror_i    (mem_rerror)
+  );
+
+  // Use the supplied clock as the XADC DCLK.
+  // This is divided internally to generate ADCCLK.
+  assign drp_dclk_o = clk_i;
+
+  assign drp_dwe_o = mem_we;
+
+  // Conversion from TLUL byte addresses to DRP register addresses
+  // is done in tlul_adapter_sram.
+  assign drp_daddr_o = mem_addr[XAw-1:0];
+
+  // Use only the bottom 16 bits of write data
+  assign drp_di_o = mem_wdata[XDw-1:0];
+  logic unused_wdata_bits;
+  assign unused_wdata_bits = ^mem_wdata[TDw-1:XDw];
+
+  // The DRP interface can process only one request at a time and can take
+  // some cycles to fulfil a request. So we need to track whether we have
+  // an outstanding read/write request the DRP is still working on.
+  logic drp_busy_d;
+  logic drp_busy_q;
+  assign drp_busy_d = drp_den_o || (drp_busy_q && ~drp_drdy_i);
+  always_ff @(posedge clk_i or negedge rst_ni) begin
+    if (!rst_ni) begin
+      drp_busy_q <= '0;
+    end else begin
+      drp_busy_q <= drp_busy_d;
+    end
+  end
+
+  // General request issue detection
+  logic above_addr_range;
+  // All accesses should be to an address offset inside the
+  // range of the 128 (0-127) XADC DRP registers.
+  assign above_addr_range = |mem_addr[TAw-1:XAw];
+
+  // Write-specific request issue detection
+  logic wmask_bad, wr_to_rd_only_addr;
+  // Writes should be to the lower 16-bits (no more & no less) in this simple
+  // implementation, as the underlying XDC registers are all 16-bit registers.
+  assign wmask_bad = mem_we && (~&mem_wmask[XDw-1:0] || |mem_wmask[TDw-1:XDw]);
```

I'd allow full 32-bit writes as well as lower 16-bit writes. In general all device memory accesses are done on aligned 32 bit accesses, so this would be surprising special case behaviour, especially as it would just silently ignore 32-bit writes.

---

Alright, I'm happy to change this if it would be a better fit for the system. It sounds like it may have been causing problems on the software side.